### PR TITLE
Withrottle steal test

### PR DESCRIPTION
**Pull request by pabender**
@BHoffman351 this PR has the tests of the steal process at the MultiThrottle level.  

The first test acquires L1234 with a steal.

The second test asks for L1234, ignores the steal prompt (rejecting the steal) and then acquires L4321 with a steal.

### CODE REVIEW

**Comment by pabender** [comment]
I will also add this test is based on the documented steal process for LNWI in 
https://github.com/JMRI/JMRI/issues/4722

File: java/test/jmri/jmrit/withrottle/MultiThrottleStealTest.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+       // set the address
+       throttle.handleMessage("+L1234<;>L1234");
+       Assert.assertFalse("Address Found",tcls.hasAddressBeenFound());
+       // the throttle manager send a steal request, which triggers a message 
+       // from the controller to the device 
+       Assert.assertEquals("outgoing message after throttle request", "MASL1234<;>L1234",cis.getLastPacket() );
+       // the device then confirms the steal.
+       throttle.handleMessage("+L1234<;>L1234");
+       // and the sequence continues as normal.
+       Assert.assertTrue("Address Found",tcls.hasAddressBeenFound());
```
Comment (BHoffman351) [inline java/test/jmri/jmrit/withrottle/MultiThrottleStealTest.java:35]: This line fails

File: java/test/jmri/jmrit/withrottle/MultiThrottleStealTest.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+       Assert.assertEquals("outgoing message after throttle request", "MASL1234<;>L1234",cis.getLastPacket() );
+       // to refuse the steal, we have to send a different address
+       throttle.handleMessage("+L4321<;>L4321");
+       Assert.assertFalse("Address Found",tcls.hasAddressBeenFound());
+       // from the controller to the device 
+       Assert.assertEquals("outgoing message after throttle request", "MASL4321<;>L4321",cis.getLastPacket() );
+       // the device then confirms the steal.
+       throttle.handleMessage("+L4321<;>L4321");
+       // and the sequence continues as normal.
+       Assert.assertTrue("Address Found",tcls.hasAddressBeenFound());
```
Comment (BHoffman351) [inline java/test/jmri/jmrit/withrottle/MultiThrottleStealTest.java:61]: This line fails for same reason as line 31.

File: java/test/jmri/jmrit/withrottle/MultiThrottleStealTest.java (hunk 1 above)
Comment (BHoffman351) [reply java/test/jmri/jmrit/withrottle/MultiThrottleStealTest.java:35]: At this point the address will be declined, the code has not done the steal request yet. The whole process is:
Request address.
Declined.
(User chooses to proceed with steal)
Request same address
Declined.
JMRI says "No Gimme!"
Address stolen

**Review by BHoffman351 — CHANGES_REQUESTED** [review]
@pabender I have to admit a lot of confusion about what is included in this PR.  The other commits that are in it seem unrelated. I have no idea if all of this should be merged into my branch or if it is already in master of JMRI/JMRI or if it will create merge conflicts.

As far as the tests, the two fail while the same process works on actual hardware.

File: java/test/jmri/jmrit/withrottle/MultiThrottleStealTest.java (hunk 1 above)
Comment (pabender) [reply java/test/jmri/jmrit/withrottle/MultiThrottleStealTest.java:35]: I thought I changed that.  Let me verify I uploaded it.

**Comment by pabender** [comment]
This PR is based on an up to date master.  If you pull the up to date master into your branch, all the extra files will go away

**Comment by pabender** [comment]
Also, the test is based on the a test throttle manager that requires the steal process be followed as designed.  It always responds to the initial request for the throttle by indicating a steal is required.  If you say no to the steal prompt, the request stops.  If you say yes, you get the throttle.

File: java/test/jmri/jmrit/withrottle/MultiThrottleStealTest.java [hunk 3]
```diff
@@ -26,7 +26,7 @@ public void testSetAndReleaseLongAddressWithSteal(){
        // from the controller to the device 
        Assert.assertEquals("outgoing message after throttle request", "MASL1234<;>L1234",cis.getLastPacket() );
        // the device then confirms the steal.
-       throttle.handleMessage("+L1234<;>L1234");
+       throttle.handleMessage("SL1234<;>L1234");
```
Comment (pabender) [inline java/test/jmri/jmrit/withrottle/MultiThrottleStealTest.java:32]: @BHoffman351 , you were looking at the wrong version of this line.  I couldn't see it in  what you quoted, but this is the correct version according to the documentation (with an S, not a +) for the steal.

File: java/test/jmri/jmrit/withrottle/MultiThrottleStealTest.java (hunk 1 above)
Comment (pabender) [reply java/test/jmri/jmrit/withrottle/MultiThrottleStealTest.java:35]: So are you telling me that WiThrottle always declines the steal the first time, and then completely re-initiates the sequence?  That doesn't match the sequence described in https://github.com/JMRI/JMRI/issues/4722.

The command sequence currently sent by this test is:
Client sends (at this level,no MA prefix...):   +L1234<;>L1234
Server Replies: MASL1234<;>L1234
Client sends: SL1234<;>L1234
Server Replies:MA+L1234<;>

in words, that is:
client requests address
server replies you have to steal it
client replies steal it
server replies you have the address.

And then there is a release after that.

**Comment by BHoffman351** [comment]
> On Nov 20, 2018, at 8:04 PM, Paul Bender <notifications@github.com> wrote:
> 
> This PR is based on an up to date master. If you pull the up to date master into your branch, all the extra files will go away

I had done that when I saw them, but the files all remained.  Some of them were from 13 days prior.

Hopefully there should be no issues with it.

**Comment by BHoffman351** [comment]
> On Nov 20, 2018, at 8:08 PM, Paul Bender <notifications@github.com> wrote:
> 
> Also, the test is based on the a test throttle manager that requires the steal process be followed as designed. It always responds to the initial request for the throttle by indicating a steal is required. If you say no to the steal prompt, the request stops. If you say yes, you get the throttle.

It seems like it might be more effective if it was based on the end result, then it would be independent of the process used to get there.

**Comment by BHoffman351** [comment]
> On Nov 20, 2018, at 8:16 PM, Paul Bender <notifications@github.com> wrote:
> 
> you were looking at the wrong version of this line. I couldn't see it in what you quoted, but this is the correct version according to the documentation (with an S, not a +) for the steal.

I think I just copied the wrong one into the comment, I did see that it was changed.  The lines that error are 31 and 54.

**Comment by BHoffman351** [comment]
> On Nov 20, 2018, at 8:34 PM, Paul Bender <notifications@github.com> wrote:
> 
> So are you telling me that WiThrottle always declines the steal the first time, and then completely re-initiates the sequence? 
> 
The WiThrottle server always declines the first with cancelThrottleRequest().  When the SL1234 comes in to the server, it flags the MultiThrottleController as being in a steal mode, requests the address again with the same exact process and when it gets notifyStealThrottleRequired(LocoAddress address) it then forces the steal with stealThrottleRequest(address, this, true).
> That doesn't match the sequence described in JMRI#4722 <https://github.com/JMRI/JMRI/issues/4722>.
> 
That is correct, it does not.  The phone/JMRI description was deduced by John using non-functional JMRI steal code.

The sequence I described is in fact the one discussed between me and AJ Ireland.  He requested that WiThrottle be able to steal so that he could implement it into the LNWI.  If it was added in such a way that required a response from a device running WiThrottle, it would leave devices that are running legacy versions of WiThrottle without a way to respond.  This would hang the module unless a timer was used.  Both he and I did not like the timer.  The cancel was agreed upon.  What he finally did as code I have no way to verify, other than to have tested WiThrottle and its ability to steal.  This was done prior to the release of the LNWI.
> The command sequence currently sent by this test is:
> 
> Client sends (at this level,no MA prefix...): +L1234<;>L1234
> Server Replies: MASL1234<;>L1234
> Client sends: SL1234<;>L1234
> Server Replies:MA+L1234<;>
> 
This is correct, but part of the process is handled completely inside the server.

> that is client requests address
> 
Declined. cancelThrottleRequest()
> 
> server replies you have to steal it
> client replies steal it
> 
Requests same address
> server replies you have to steal it - This causes the failure of the test
> 
Server then sends steal command
> server replies you have the address.
> 


And that would work, but it is missing the step in the middle.  If the test waited for one more action to complete, it would pass.  As is, it is checking for the address to have been acquired when the ThrottleManager is responding that the address is in use.  (It is still false at that point.)  The MultiThrottleController will then send the stealThrottleRequest(address, this, true) and actually get the address. (Now it will be true.)
> And then there is a release after that.
> 
Which should be fine.

**Comment by pabender** [comment]
> On Nov 20, 2018, at 8:08 PM, Paul Bender ***@***.***> wrote: Also, the test is based on the a test throttle manager that requires the steal process be followed as designed. It always responds to the initial request for the throttle by indicating a steal is required. If you say no to the steal prompt, the request stops. If you say yes, you get the throttle.
> It seems like it might be more effective if it was based on the end result, then it would be independent of the process used to get there.

That may be, but to get to the end result, you have to follow the right communication sequence with the ThrottleManager.  Using this version of the ThrottleManager ensures that happens.  The same ThrottleManager used for this test is also used for the steal tests in the JMRI throttle.

The test is basically treating the MultiThrottleCode like a black box.  

I'm not really concerned about the interaction between the class under test here and the client, other than we're using the commands from the client to interact with the server.  If the commands in the test sequence don't match what the client is sending, then we'll fix that, but if they do, then there is something not right in the.  So you tell me, what is the correct sequence, and I'll update the tests.

**Comment by pabender** [comment]
> that is client requests address
> Declined. cancelThrottleRequest()

Ok, I found it.  that happens outside of both MultiThrottle and MultiThrottleController.... I think I have that step included now.

File: java/test/jmri/jmrit/withrottle/MultiThrottleStealTest.java (hunk 3 above)
Comment (BHoffman351) [reply java/test/jmri/jmrit/withrottle/MultiThrottleStealTest.java:32]: Yes, That is correct.

**Review by BHoffman351 — APPROVED** [review]

**Pull request merged**